streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -577,8 +555,7 @@ void removeAllSensors() {
             removeSensor(punctuateTimeSensor);
             removeSensor(taskCreatedSensor);
             removeSensor(tasksClosedSensor);
-            removeSensor(skippedRecordsSensor);
-
+            removeSensor(skippedRecordsSensor());
```

Comment:
Hmm.. why we create the sensor in `StreamsMetricsmpl` while removing it in `StreamsMetricsThreadImpl`? It seems a bit inconsistency.. could we still create in `StreamsMetricsThreadImpl`, and let `StreamsMetricsImpl` to get a hold of the sensor object assuming it is already created then (i.e. set `sensor == null` in constructor, and in `skippedRecordsSensor(): if (sensor == null) try get it from the metrics`)?

Reply:
I tried that, but it wound up making things messier than I expected (because there are other callers to StreamsMetricsImpl, and because it makes the ownership of this sensor ambiguous).

Instead, I added https://github.com/apache/kafka/pull/4812/commits/36c065fa93b922668af87c8c7cc00460a7282ced, which gives SMI the ability to remove its own sensors, and then I called to it from the two places (StreamThread and GlobalStreamThread) that actually need to unload metrics when they shut down.

WDYT?

Reply:
I did not completely get your explanation re: `because there are other callers to StreamsMetricsImpl, and because it makes the ownership of this sensor ambiguous`. Could you elaborate a bit more?

Reply:
Sorry; I misread your suggestion. I thought you wanted the StreamsMetricsImpl to take the sensor as a constructor argument. 

Aside from the StreamThread (via StreamThreadMetricsImpl), several other classes directly invoke the StreamsMetricsImpl constructor and thus obtain a StreamsMetricsImpl that is not a StreamThreadMetricsImpl. Namely, GlobalStreamThread, MockProcessorContext, and TopologyTestDriver.

When the code paths downstream of these points need to record a skipped record, they will get a null sensor back. It wouldn't be possible to get it from the Metrics registry at that point, though, because the skipped records sensor is scoped by thread (or "virtual" thread for the test-utils), and the sensor would never have been created for GlobalStreamThread, MockProcessorContext, or TopologyTestDriver. So the only way to get it at that point is to have either the caller of the SMI constructor or the SMI itself create the sensor (either at construction or at call-time).

The previous implementation with the public static getter was effectively saying that the one who wants it in a particular context first creates it, but it's problematic because no-one owns it. And indeed, in my implementation, the sensor never got destroyed. In practice I think it's not a huge deal because I'm sure it's rare for a streams app to shut down and start up again with the same Metrics registry, and I think the threads live as long as the app. But still, we have this model of unloading metrics when they're out of scope, and I think it's a good one.

So that brings us to the current implementation. In the current implementation, the skipped records metric is clearly owned by the StreamsMetricsImpl, which it may as well be, since that is the smallest scope in which it's needed. It's the first metric to be owned by SMI, so I had to create a removeAll method, and make sure it's invoked in the right places. But that seems appropriate; every other scope that owns metrics has such a method.